transport/RtcpReportProducer.h

```diff
@@ -0,0 +1,59 @@
+#pragma once
+
+#include "concurrency/MpmcHashmap.h"
+#include "config/Config.h"
+#include "memory/Packet.h"
+#include "memory/PacketPoolAllocator.h"
+#include "transport/RtpReceiveState.h"
+#include "transport/RtpSenderState.h"
+#include "utils/Optional.h"
+#include "utils/Span.h"
+
+namespace transport
+{
+
+class RtcpReportProducer
+{
+public:
+    struct RtcpSender
+    {
+        virtual ~RtcpSender() = default;
+        virtual void sendRtcp(memory::UniquePacket packet, uint64_t timestamp) = 0;
+    };
+
+    RtcpReportProducer(const logger::LoggableId& loggableId,
+        const config::Config& config,
+        const concurrency::MpmcHashmap32<uint32_t, RtpSenderState>& outboundSsrcCounters,
+        const concurrency::MpmcHashmap32<uint32_t, RtpReceiveState>& inboundSsrcCounters,
+        memory::PacketPoolAllocator& rtcpPacketAllocator,
+        RtcpSender& rtcpSender);
```

Comment:
I would like to avoid these callback loops when possible. In several other similar situations we ask the "factory" repeatedly for data until it returns false or null. The factory can maintain state easily in most cases.

It makes it a little bit more decoupled and easier to follow call stacks when they are more of a tree than a graph.

Reply:
I can do that but I think it will cause more complexity in this case. Because we have to keep the state of ReportContext between multiple calls to generate the report. Moreover, we use stack memory to store the ssrc to send. That change would require a big change in the implementation of RtcReportProducer